### PR TITLE
node: metrics: add metrics about cpu pool sizes

### PR DESCRIPTION
**Pull request by ffromani**
/kind feature

#### What this PR does / why we need it:
Add metrics about cpu pools size. Givens more insight about how resources are allocated to workload and how node resources are used, and enables future changes (xref: https://github.com/kubernetes/enhancements/pull/4541#issuecomment-2363006425)

This work includes a fix from #120661 . We are bundling the fix in this PR because of the interdependency between thre new metrics, its test and the fix needed by the test to pass. See: https://github.com/kubernetes/kubernetes/pull/127506#issuecomment-2388198994 and https://github.com/kubernetes/kubernetes/pull/120661#issuecomment-2387765036 and https://github.com/kubernetes/kubernetes/pull/120661#issuecomment-2388198443

#### Which issue(s) this PR fixes:
Fixes #N/A

#### Special notes for your reviewer:
The shared pool size is pretty much easy to represent, but do we want more granularity about the exclusive cpu assignments? histogram perhaps? Implemented the simplest approach so far.

#### Does this PR introduce a user-facing change?
```release-note
Added a kubelet metrics to report informations about the cpu pools managed by cpumanager when the static policy is in use.
```

### CODE REVIEW

**Comment by ffromani** [comment]
/hold

similarly to https://github.com/kubernetes/kubernetes/pull/127155 I'd like to add some tests. Besides that the PR is fully reviewable

**Comment by ffromani** [comment]
note a test I added is known broken because https://github.com/kubernetes/kubernetes/pull/120661is not merged
I can add a workaround or we can merge 120661

**Comment by ffromani** [comment]
/retest

```
Kubernetes e2e suite: [It] [sig-node] Kubelet when scheduling an agnhost Pod with hostAliases should write entries to /etc/hosts [NodeConformance] [Conformance] expand_more
```
very unlikely this is this PR's fault

**Comment by ffromani** [comment]
ok, good news: #120661 fixes the test I added in https://github.com/kubernetes/kubernetes/pull/127506/files#diff-897688174d4d6c5ec8b4080a2bbeeddc9e5442781586de6154a32b0b80cfc336R199 , hence the latter (or an amended version of) can be used to validate the former

**Comment by ffromani** [comment]
/hold

depends on https://github.com/kubernetes/kubernetes/pull/120661 (see: https://github.com/kubernetes/kubernetes/pull/120661#issuecomment-2388313796)

**Comment by ffromani** [comment]
@SergeyKanzhelev @haircommander (cc @likakuli )

ready for review!
xref: https://github.com/kubernetes/kubernetes/pull/120661#issuecomment-2399530498

**Comment by haircommander** [comment]
should we change the desription of this to say it's also doing https://github.com/kubernetes/kubernetes/pull/120661 or should we merge https://github.com/kubernetes/kubernetes/pull/120661 before merging this?

**Comment by ffromani** [comment]
> should we change the desription of this to say it's also doing #120661 or should we merge #120661 before merging this?

good point, fixed the PR description

**Comment by ffromani** [comment]
serial-crio-cgroupv* lane failures are unrelated. Other lanes e.g. serial-cpu-manager exercise this flow correctly.

File: test/e2e_node/podresources_test.go [hunk 1]
```diff
@@ -740,21 +737,30 @@ func podresourcesGetAllocatableResourcesTests(ctx context.Context, cli kubeletpo
...
-	gomega.ExpectWithOffset(1, devs).ToNot(gomega.BeEmpty())
+	gomega.Expect(devs).ToNot(gomega.BeEmpty())
 	for _, dev := range devs {
 		gomega.Expect(dev.ResourceName).To(gomega.Equal(sd.resourceName))
-		gomega.ExpectWithOffset(1, dev.DeviceIds).ToNot(gomega.BeEmpty())
+		gomega.Expect(dev.DeviceIds).ToNot(gomega.BeEmpty())
+	}
+}
+
+func allocatableCPUsFromGetAllocatableResourcesRequest(resp *kubeletpodresourcesv1.AllocatableResourcesResponse) (cpuset.CPUSet, []*kubeletpodresourcesv1.ContainerDevices) {
```
Comment (likakuli) [inline test/e2e_node/podresources_test.go:757]: nit: The function name seems a bit odd; from the results, it returns both CPU and device information, but the name specifically mentions only the CPU

File: test/e2e_node/cpu_manager_metrics_test.go [hunk 2]
```diff
@@ -189,7 +282,7 @@ func makeGuaranteedCPUExclusiveSleeperPod(name string, cpus int) *v1.Pod {
 							v1.ResourceMemory: resource.MustParse("64Mi"),
 						},
 					},
-					Command: []string{"sh", "-c", "sleep", "1d"},
+					Command: []string{"sh", "-c", "sleep 1d"},
```
Comment (likakuli) [inline test/e2e_node/cpu_manager_metrics_test.go:285]: nice catch 👍

File: test/e2e_node/podresources_test.go (hunk 1 above)
Comment (ffromani) [reply test/e2e_node/podresources_test.go]: thanks, addressed!

**Comment by ffromani** [comment]
@SergeyKanzhelev @haircommander rebased and reviewable. This is the second and last PR adding metrics for 1.32 that I planned.

File: pkg/kubelet/metrics/metrics.go [hunk 3]
```diff
@@ -773,6 +775,27 @@ var (
 		},
 	)
 
+	// CPUManagerSharedPoolSizeMilliCores reports the current size of the shared CPU pool for non-guaranteed pods
+	CPUManagerSharedPoolSizeMilliCores = metrics.NewGauge(
+		&metrics.GaugeOpts{
+			Subsystem:      KubeletSubsystem,
+			Name:           CPUManagerSharedPoolSizeMilliCoresKey,
+			Help:           "The size of the shared CPU pool for non-guaranteed QoS pods, millicores.",
```
Comment (haircommander) [inline pkg/kubelet/metrics/metrics.go:783]: nit: `, millicores` doesn't read right grammatically, and it's not that much longer to write `in millicores`

File: pkg/kubelet/metrics/metrics.go [hunk 4]
```diff
@@ -773,6 +775,27 @@ var (
...
+			StabilityLevel: metrics.ALPHA,
+		},
+	)
+
+	// CPUManagerExclusiveCPUsAllocationCount reports the total number of CPUs exclusively allocates to containers running on this node
+	CPUManagerExclusiveCPUsAllocationCount = metrics.NewGauge(
+		&metrics.GaugeOpts{
+			Subsystem:      KubeletSubsystem,
+			Name:           CPUManagerExclusiveCPUsAllocationCountKey,
+			Help:           "The total number of CPUs exclusively allocates to containers running on this node",
```
Comment (haircommander) [inline pkg/kubelet/metrics/metrics.go:793]: I think `allocated` would read better here (and in the comment above)

File: pkg/kubelet/metrics/metrics.go (hunk 3 above)
Comment (ffromani) [reply pkg/kubelet/metrics/metrics.go:783]: just me being sloppy in english ;) fixed, thanks

File: pkg/kubelet/metrics/metrics.go (hunk 4 above)
Comment (ffromani) [reply pkg/kubelet/metrics/metrics.go:793]: Good catch, this is just a typo. Fixed.

File: pkg/kubelet/metrics/metrics.go (hunk 3 above)
Comment (haircommander) [reply pkg/kubelet/metrics/metrics.go]: since I only know one language, I might as well be annoying about it :)

**Comment by ffromani** [comment]
/triage accepted

We usually don't self-triage but this seems kinda evident from the context

**Comment by ffromani** [comment]
ping (dep for https://github.com/kubernetes/kubernetes/pull/127483 even though the two PRs can be merged in any order)

**Comment by ffromani** [comment]
/priority important-longterm

> /triage accepted
> 
> We usually don't self-triage but this seems kinda evident from the context

likewise for priority (prerequisite for an accepted KEP: https://github.com/kubernetes/kubernetes/issues/122955)

**Pull request merged**